gomods.go

```diff
@@ -0,0 +1,253 @@
+}
+
+type ModuleHandler interface {
+	fetch(r *http.Request, c Config) (*storage.Version, error)
+	storage(c Config) (storage.Backend, error)
+	dp(fetcher module.Fetcher, s storage.Backend, fs afero.Fs) download.Protocol
+}
+
+var gomodsRegex = regexp.MustCompile("(list|info|mod|zip)")
+var modVersionRegex = regexp.MustCompile("@v\\/(v\\d+\\.\\d+\\.\\d+\\-\\d+\\-[\\w\\d]+|v\\d+\\.\\d+\\.\\d+|v\\d+\\.\\d+|latest|master)")
+var DefaultGoBinaryPath = os.Getenv("GOROOT") + "/bin/go"
+
+const (
+	DefaultGomodsCacheType = "tmp"
+	DefaultGomodsWorkers   = 1
+)
+
+// SetDefaults sets the default values for gomods config
+// if the fields are empty
+func (gomods *Gomods) SetDefaults() {
+	gomods.Fs = afero.NewOsFs()
+	if gomods.GoBinary == "" {
+		gomods.GoBinary = DefaultGoBinaryPath
+	}
+	if gomods.Cache.Enable {
+		if gomods.Cache.Type == "" {
+			gomods.Cache.Type = DefaultGomodsCacheType
+		}
+		if gomods.Cache.Path == "" {
+			gomods.Cache.Path = afero.GetTempDir(gomods.Fs, "")
+		}
+	}
+	if gomods.Workers == 0 {
+		gomods.Workers = DefaultGomodsWorkers
+	}
+}
+
+func gomods(w http.ResponseWriter, r *http.Request, path string, c Config) error {
+	moduleName, version, ext := moduleNameAndVersion(path)
+	if moduleName == "" {
+		return fmt.Errorf("module url is empty")
+	}
+	m := Module{
+		Name:    moduleName,
+		Version: version,
+	}
+	dp, err := m.fetch(r, c)
+	if err != nil {
+		return err
+	}
+	switch ext {
+	case "list":
+		list, err := dp.List(r.Context(), m.Name)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write([]byte(strings.Join(list, "\n")))
+		if err != nil {
+			return err
+		}
+		return nil
+	case "info":
+		info, err := dp.Info(r.Context(), m.Name, m.Version)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(info)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "mod":
+		mod, err := dp.GoMod(r.Context(), m.Name, m.Version)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(mod)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "zip":
+		zip, err := dp.Zip(r.Context(), m.Name, m.Version)
+		if err != nil {
+			return err
+		}
+		defer zip.Close()
+		w.Write([]byte{})
+		_, err = io.Copy(w, zip)
+		if err != nil {
+			return err
+		}
+		return nil
+	case "latest":
+		info, err := dp.Latest(r.Context(), m.Name)
+		if err != nil {
+			return err
+		}
+		json, err := json.Marshal(info)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(json)
+		if err != nil {
+			return err
+		}
+		return nil
+	default:
+		return fmt.Errorf("the requested file's extension is not supported")
+	}
+}
+
+func (m Module) fetch(r *http.Request, c Config) (download.Protocol, error) {
+	fetcher, err := module.NewGoGetFetcher(c.Gomods.GoBinary, c.Gomods.Fs)
+	if err != nil {
+		return nil, err
+	}
+	s, err := m.storage(c)
+	if err != nil {
+		return nil, err
+	}
+	dp := m.dp(fetcher, s, c)
+	return dp, nil
+}
+
+func (m Module) storage(c Config) (storage.Backend, error) {
+	switch c.Gomods.Cache.Type {
```

Comment:
Shouldn't "tmp" be handled here as well? Or am I missing something?

Reply:
Ok we are setting the tmp path/fs within the defaulting section. This feels a bit weird, but I see the reasoning.